Add NavigationPathQuery objects and NavigationServer query_path()

smix8 opened the pull request:
Adds `NavigationPathQueryParameters` and `NavigationPathQueryResult` objects in 2D and 3D similar to e.g. physics query parameters and results.

Can be used with 2D or 3D NavigationServer `query_path()` to get a customized navigation path considering additional options.

Does not affect the normal pathfinding with NavigationAgent nodes or NavigationServer.map_get_path() at all. Without customized options returns the same path as map_get_path() for now.

Removed for this pr most other options that are still developed and not finished yet (like [#2527](https://github.com/godotengine/godot-proposals/issues/2527) or [#3812](https://github.com/godotengine/godot-proposals/issues/3812)). If there is only a single function or an enum has only a single value that is why. This is basically only the core left as many other pr's and features will depend on this. This query can be used as a base to fix issues #19011 and #60277 in the future by adding a property that controls how the pathfinding adds points or a simplifying method in path postprocessing.

This pr is also a prerequisite for #62115 or any other issue that is related to the currently used pathfinding algorithm and forced path post-processing. Since Godot seems that it wants to support both 2D and 3D pathfinding as free-form as well as tile / grid based with TileMap and GridMap (#61293) it is impossible to have just one pathfinding algorithm and one out of two forced path post-processing for everything like it is rightnow.

Usage is to create both objects first and store them into a var for reuse.
Then use query_path() for updating the content of the result.

Example:

```gdscript
var query_parameters : NavigationPathQueryParameters3D = NavigationPathQueryParameters3D.new()
var query_result  : NavigationPathQueryResult3D = NavigationPathQueryResult3D.new()

# fill in parameters...

NavigationServer3D.query_path(query_parameters, query_result)
var path = query_result.get_path()
```
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

smix8 committed:
Add NavigationPathQuery

Adds NavigationPathQueryParameters objects that can be used with NavigationServer.query_path() to query a customized navigation path.

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+			Vector3 path_pos_next = p_query_result.path[path_inx + 1];
+
+			if (path_pos == path_pos_next) {
+				path_inx += 1;
```
Should this have a tolerance?

smix8 replied:
This was more for the cases where pathfinding would add the same point multiple times due to the navmesh layout so the check would fail if points overlap completely.

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+
+				if (dotp > margin_limit) {
+					is_straight = true;
+				}
```
Is dot product the best way to do this? Consider that over large distances the deviation will be larger from a path than with smaller distances. Perhaps you can take this into account in the calculation.

AndreaCatania replied:
I wonder what `margin_limit` value is. Is it an angle? Because if it's an angle in radiant you need to convert the dot product to an angle using `Math::acos(dotp)`.

smix8 replied:
Not sure if dot product is the best way, it is the way I figured it out with my C-fu as I am not aware of a "better" way. Margin is radian. Maybe for usability this should be handled like with 3d nodes where the user visible property is in angle but the internals are in radian to avoid converting it.

smix8 replied:
The path simplifying was removed.

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+}
+
+void GodotNavigationServer::_strip_path(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	Vector<Vector3> stripped_path;
```
"_strip_path" is a bit vague, maybe "simplify path" would be better?

smix8 replied:
For me it was "strip path from unnecessary points" but "simplify" sounds better.

lawnjelly commented on servers/navigation_server_2d.cpp:
```diff
@@ -305,3 +307,170 @@ bool FORWARD_1_C(agent_is_map_changed, RID, p_agent, rid_to_rid);
+
+	for (int i = 0; i < _query_result.path.size(); i++) {
+		path_vec2.push_back(Vector2(_query_result.path[i].x, _query_result.path[i].z));
+	}
```
Generally at the moment it is better to resize `path_vec2` in one call if you know the size in advance, then set each element rather than use `push_back`, to avoid multiple allocations.

lawnjelly commented on servers/navigation_server_2d.cpp:
```diff
@@ -305,3 +307,170 @@ bool FORWARD_1_C(agent_is_map_changed, RID, p_agent, rid_to_rid);
+	}
+
+	Ref<NavigationPathQueryResult2D> query_result(memnew(NavigationPathQueryResult2D));
+
```
Do we need to use dynamic allocation here? Can't we just pass a `NavigationPathQueryResult2D` by reference and have the function fill it? That way it is up to the caller whether to use dynamic allocation or stack etc.

smix8 replied:
No dynamic is not needed here at the moment. Should be indeed a reuseable reference that users create once like the parameters and send with every query_path() for only updating the values. The dynamic alloc makes the path request very slow in my tests compared to the default map_get_path() so it wouln't scale well when updated every frame with many actors.

lawnjelly commented on servers/navigation_server_2d.cpp:
```diff
@@ -305,3 +307,170 @@ bool FORWARD_1_C(agent_is_map_changed, RID, p_agent, rid_to_rid);
+		default: {
+			parameters.pathfinding_algorithm = gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR;
+		} break;
+	}
```
Do there need to be two different enums here? Is it possible we can just set one equal to the other?

smix8 replied:
I split that because it is entirely possible that we have different enum values between 2D and 3D and those enums are exposes in docs for users.

lawnjelly commented on servers/navigation_server_2d.cpp:
```diff
@@ -305,3 +307,170 @@ bool FORWARD_1_C(agent_is_map_changed, RID, p_agent, rid_to_rid);
+
+void NavigationPathQueryParameters2D::set_navigation_layers(uint32_t p_navigation_layers) {
+	parameters.navigation_layers = p_navigation_layers;
+};
```
Personally, if these are one liners, I'd put them in the header, just to make things less verbose. But it depends if you think you are likely to add to the function implementations later.

smix8 replied:
I think at some point all the setters could require some dirty flags to trigger update signals. I also have to admit I simply prefer this to one liners for my codeditor usage.

lawnjelly commented on doc/classes/NavigationPathQueryParameters2D.xml:
```diff
@@ -0,0 +1,45 @@
+		</member>
+		<member name="strip_margin" type="float" setter="set_strip_margin" getter="get_strip_margin" default="0.001">
+			Margin value for normalized vector dot product to decide if intermediate path positions form enough of a straight line to be stripped. Used if [member strip_path] is [code]true[/code].
+		</member>
```
This naming is not great for a dot product. But I suspect the name "margin" could be kept, but it should actually be a distance in world space for deviation from the path.

Although you may want to also have a param for tolerance of equal points so if this is not hard coded, these two names should make sense in terms of each other.

smix8 replied:
Naming is not great, I actuall don't know how to describe that better, and suggestion?

DarkKilauea replied:
Maybe try `path_angular_deviation`?

smix8 replied:
The path simplifying was removed.

lawnjelly commented:
Have written some comments. But really before adding this PR I think the navigation team should decide whether they are going to add async, and if so add this first because the API and implementation of all this stuff may need to be different if supporting async pathfinding.

AndreaCatania commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+		if (path_inx + 1 < p_query_result.path.size()) {
+			Vector3 path_pos_next = p_query_result.path[path_inx + 1];
+
+			if (path_pos == path_pos_next) {
```
I think here you should use `Math::is_equal_approx` https://github.com/godotengine/godot/blob/master/core/math/vector3.h#L141 or `(path_pos - path_pos_next).length_squared() <= tollerance`

AndreaCatania commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+	p_query_result.path = stripped_path;
+}
+
+void GodotNavigationServer::_calc_path_length(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
```
`p_query_result` should be `r_query_result` instead.

smix8 replied:
How? Shouldn't all Godot function parameters be prefixed with `p_`?. `p_query_result` is from the function PathQueryResult parameter, there is no return.

AndreaCatania commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+		Vector3 prev_path_pos = p_query_result.path[0];
+		for (int i = 0; i < p_query_result.path.size(); i++) {
+			Vector3 path_pos = p_query_result.path[i];
+			Vector3 prev_pos_to_path_pos = path_pos - prev_path_pos;
```
Also here consider using `const Vector3 &` for `prev_path_pos` and `path_pos`

AndreaCatania commented on modules/navigation/godot_navigation_server.h:
```diff
@@ -153,6 +153,12 @@ class GodotNavigationServer : public NavigationServer3D {
+
+private:
+	void _strip_path(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const;
+	void _calc_path_length(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const;
```
I'm not entirely sure if this is a new convention in godot, but I think you can avoid prefixing with `_` even when the function is private. In other context prefix with `_` is used to signal that such thing is not used, also it seems just useless prefix the private function names as those are not used externally anyway.
However @akien-mga knows better than I do.

akien-mga replied:
No strong opinion here, there's a bit of everything in the codebase. `_` prefix for private methods is not uncommon and to me it makes sense. But it's also ok without.

smix8 replied:
Personal preference to have helper functions that see only internal use with a underscore prefix. If it is no issue I would just keep it.

AndreaCatania reviewed (commented):
It looks good, just few nits!

Scony commented on doc/classes/NavigationPathQueryParameters2D.xml:
```diff
@@ -0,0 +1,45 @@
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="calc_path_length" type="bool" setter="set_calc_path_length" getter="get_calc_path_length" default="false">
```
how about changing `calc` to `calculate`

Scony commented on doc/classes/NavigationPathQueryParameters3D.xml:
```diff
@@ -0,0 +1,45 @@
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="calc_path_length" type="bool" setter="set_calc_path_length" getter="get_calc_path_length" default="false">
```
same

Scony commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+void GodotNavigationServer::_strip_path(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	Vector<Vector3> stripped_path;
+
+	int path_inx = 0;
```
how about `path_index`

Scony commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -635,6 +635,99 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+	Vector3 past_direction;
+	bool has_past_direction = false;
+	float margin_limit = 1.0 - p_parameters.strip_margin;
+	float dotp = 0.0;
```
how about `dot_product`

Scony commented on servers/navigation_server_3d.cpp:
```diff
@@ -121,3 +123,164 @@ NavigationServer3D *NavigationServer3DManager::new_default_server() {
+			parameters.pathfinding_algorithm = gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR;
+		} break;
+		default: {
+			parameters.pathfinding_algorithm = gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR;
```
I guess we can hit default iff new, unhandled algorithm was added. IMO it's worth putting warning about "not implemented" case here

smix8 replied:
Right, I guess a WARN_PRINT_ONCE here might not hurt in case default is hit.

Scony commented on servers/navigation_server_3d.cpp:
```diff
@@ -121,3 +123,164 @@ NavigationServer3D *NavigationServer3DManager::new_default_server() {
+		case gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR:
+			return PATHFINDING_ALGORITHM_ASTAR;
+		default:
+			return PATHFINDING_ALGORITHM_ASTAR;
```
same here

DarkKilauea commented on modules/navigation/navigation_path_query_parameters_2d.h:
```diff
@@ -0,0 +1,86 @@
+#include "core/object/ref_counted.h"
+#include "nav_utils.h"
+
+class NavigationPathQueryParameters2D : public RefCounted {
```
Should the query/result objects be part of the navigation module?  What if someone wanted to replace the module with another implementation of navigation, would they need to update the servers as well?

smix8 replied:
Godot should be able to build without the navigation module which includes all this. I don't want to clutter the resource folder with stuff that is very module specific and also didn't want to add everything to the NavigationServer source code.

At the moment Godot is not even able to be build without the navigation module so I wouldn't bother about this until we reach that point where custom servers even become a real thing.

DarkKilauea commented on modules/navigation/register_types.cpp:
```diff
@@ -53,6 +61,12 @@ NavigationServer3D *new_server() {
 
 void initialize_navigation_module(ModuleInitializationLevel p_level) {
 	if (p_level == MODULE_INITIALIZATION_LEVEL_SERVERS) {
+#ifdef MODULE_NAVIGATION_ENABLED
```
Are these guards actually useful?  I thought the entire module wouldn't be compiled if the module was disabled.

smix8 replied:
The guards are only useful to shutup the "Minimal Template" pr check that tries a build without navigation which is not possible at the moment.

DarkKilauea commented on modules/navigation/navigation_path_query_result_3d.cpp:
```diff
@@ -0,0 +1,54 @@
+	ClassDB::bind_method(D_METHOD("get_path"), &NavigationPathQueryResult3D::get_path);
+	ClassDB::bind_method(D_METHOD("get_path_length"), &NavigationPathQueryResult3D::get_path_length);
+
+	ADD_SIGNAL(MethodInfo("result_changed"));
```
It seems a little odd to me to register for a signal on a temporary result object.  How would this look in GDScript?

Ideally, I'd like to do something like:

``` gdscript

var query := NavigationPathQueryParameters3D();
# query.target_position = ...
# fill in parameters...

var result := await NavigationServer3D.query_path(query);

```

Do we know if GDScript has the concept of a `Promise` or `Awaitable` instead of only waiting for a signal?

smix8 replied:
The result is no longer a throw-away object. The idea is to create both objects and reuse them. The signal is basically there so users can connect for updates in preperation for async path queries. Since this is not added now it is not so useful cause as soon as a user uses the query_path() function the result is updated.

DarkKilauea replied:
If these objects are intended to stick around, maybe we should just have one of them to avoid confusion with the physics APIs (which look similar but don't work the same way).  It also simplifies the API by not requiring the user to keep two objects around and remember to always pair them with each other.

So, we'd have `NavigationPathQuery3D` which contains the properties of both `NavigationPathQueryParameters3D` and `NavigationPathQueryResult3D`.

Then you'd call it like:

``` gdscript
var query := NavigationPathQuery3D();

# fill in parameters...
query.target_location = Vector3(...);
query.current_location = self.global_position;

# run the query
NavigationServer3D.query_path(query);

# wait for the query to complete
await query.completed;

# get back the final path
var path = query.get_path();
```

smix8 replied:
Physics has no result object reuse and it is actually quite wasteful e.g. new object with each query or using dictionary.

The split makes sense cause you do not always want both objects created at an equal count. E.g. I have an actor manager that reuses the same parameter object by just updating locations in a timesliced loop or I have multiple parameters for gameplay situations on my actor but it only uses one result for the pathfinding.

lawnjelly replied:
Agree with @DarkKilauea on this being odd to have the signal as part of the query. Could this be a signal on the `NavigationAgent` like `velocity_computed`?
Given that this is now adapted to work in an async framework, can the signal return the result (separate from the query)?

I don't know if there are potentials for race conditions of several queries before the first is returned.

smix8 replied:
The query needs to work without any NavigationAgent node involved so where else to place a signal? Can copy the way how avoidance callbacks solve this but those use no signal and solve this by registering callbacks first and then calling the function on the object directly. The NavigationAgent emits the velocity_computed after the internal avoidance callback functions gets called.

lawnjelly replied:
This sounds fine imo, I would just copy the technique from avoidance callbacks. :+1:

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -639,6 +639,98 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+
+	p_query_result.path = simplified_path;
+}
+
```
It would be nice to see some graphical examples of paths where this is a problem currently. I'm slightly concerned that using dot product in this way could cause agents to get stuck in some situations, while also not being super intuitive for users.

Am suspecting it may be better here to wait on the ray tracing functionality for the navmesh, because that will enable auto-simplification of paths without cutting corners etc, and won't need parameterization by the user. Or else use a very conservative hard coded dot product (or similar) here as a placeholder and an on switch for path simplification.

But unless this is super pressing, it may be better to wait on this.

smix8 replied:
Will split and wait, the path simplifying is not of core importance for this pr and can be added later in fashion.

lawnjelly commented:
Here is an illustration of a problem of using the dot product to detect straight regions. A and B the dot product in each case is similar as the vectors are normalized. However the absolute error (the blue line) is huge when the distances between the waypoints is higher, which can result in agent walking into impassable area, getting stuck etc.

![dot_product](https://user-images.githubusercontent.com/21999379/184697267-000344c4-a4f8-4db8-b47e-c9dce7fb7d31.png)

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+#ifdef MODULE_NAVIGATION_ENABLED
+	_process_path_queries();
+#endif // MODULE_NAVIGATION_ENABLED
+}
```
I would add a comment or preferably rename `GodotNavigationServer::process()` to make it clear it is being done on a physics tick, not a frame. The name `process` in godot is associated with frame processing (rightly or wrongly), so this would be good to not have confused. Perhaps `physics_process()`, to be consistent with other methods.

smix8 replied:
I agree with the process() but that raises the general problem that the NavigationServer uses a function called `process()` for its main iteration that is only called with the physics_process tick.

lawnjelly replied:
Presumably it's only called from the engine internally, so should be fine to change. But it's no biggy if the main API function doesn't have this name, but I would suggest either using `physics_process` as much as possible, instead of `process`, or use another term like `update`, which is less confusing.

smix8 replied:
Maybe a general term like update is better cause if it is no longer the physics tick responsible for this function it would be confusing and require renames.

lawnjelly replied:
Yup `update` is fine, and possibly better.

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+		// NavigationPathQueryResult3D
+	} else if (Object::cast_to<NavigationPathQueryResult3D>(queryresult_object) != nullptr) {
+		NavigationPathQueryResult3D *pathqueryresult3d = Object::cast_to<NavigationPathQueryResult3D>(queryresult_object);
+
```
In both of these cases, it might be better to do the cast once to a variable, then test the variable in the if. You don't know how expensive the `Object::cast_to()` call is (it looks like it reverts to RTTI in some situations).

smix8 replied:
Agree, will move the cast for the two potential results before the if block.

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+void GodotNavigationServer::_calculate_path_length(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	float _sum = 0.0;
+	if (p_query_result.path.size() > 1) {
+		for (int i = 1; i < p_query_result.path.size(); i++) {
```
There's no need for the if statement.

smix8 replied:
Will remove

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+}
+
+void GodotNavigationServer::_calculate_path_length(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	float _sum = 0.0;
```
Should these be `real_t`?

smix8 replied:
Not sure if it should but I guess it could.

lawnjelly replied:
It's going to be calculated as 64 bit anyway if `Vector3` is 64 bit, so storing as 32 bit would be throwing away data in this case. I think it was stored as `real_t` elsewhere (in the query result or something)?

smix8 replied:
real_t it is then

lawnjelly commented on modules/navigation/nav_utils.h:
```diff
@@ -150,6 +150,34 @@ struct ClosestPointQueryResult {
+};
+
+struct PathQueryParameters {
+	PathfindingAlgorithm pathfinding_algorithm = PATHFINDING_ALGORITHM_ASTAR;
```
Any need to put this in in this PR if there is currently only one pathfind method?

lawnjelly replied:
As discussed on rocket chat (better to keep on the PR for history purposes) : 

> About the enum / pathfinding algorithms included in the pr. I included them because I think if multiple ppl start working on different algo we may end up with x-implementations all over the place. This way everyone knows the access point to add the new stuff. E.g. if Acheron and I start working on two different things we would just add our algos to the existing enum. This was the general idea why this was added while we have no different pathfinding implemented.

That seems fair, and is up to you, not something that is a show stopper. :slightly_smiling_face: 

I just try where possible to refer to the guidelines, and being wary of the section on futureproofing:
https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html

Personally I also tend to err on the side of future proofing, but try and hold back a bit for Godot because of the different philosophy. :slightly_smiling_face:

smix8 replied:
Enum and this was included so if multiple ppl start work on new algorithms they would know where to add them instead of ending up with x-different implementations all over the place. I know DarkKilauea is just waiting for this pr to be merged to start adding one better algo for the navigation jumplinks and I also want to start work, might be others too.

lawnjelly commented on modules/navigation/nav_utils.h:
```diff
@@ -150,6 +150,34 @@ struct ClosestPointQueryResult {
+
+struct PathQueryTask {
+	PathQueryParameters query_parameters;
+	ObjectID result_object_id;
```
What's the pros and cons of storing an object for the calling actor (`result_object_id`) _AND_ the receiving object in the `PathQueryParameters` (`callback_object_id`)? Maybe the callback object will be deleted if the actor is?

lawnjelly replied:
Well here it seems like the crucial thing is the `callback_object_id` rather than the actor. If the `callback_object_id` is valid, you can go ahead, if not you can't. I don't think sending / storing / checking the actor ID is necessary if all the communication is via this object.

smix8 replied:
Forget what I said about mentioned delete issues. Changed the query stuff from objects to RefCounted some weeks ago so it does not suffer the same problems as the agents had e.g. from an early user `free()` or `queue_free()` while still in the waiting room.

The `result_object_id` and `callback_object_id` are also split so callbacks can happen on objects that the users needs them, not fixated to the result, e.g. while the result is updated and used by the actor the callback that the result has been updated is received on another method or the method of an entirely different object e.g. a manager object that controls the ai actors.

The parameter `callback_object_id` is also optional in the parameters. If no `callback_object_id` or `callback_method` the query is processed immediately as was requested in devchat.

smix8 replied:
removed from this pr

lawnjelly commented on modules/navigation/navigation_path_query_parameters_2d.cpp:
```diff
@@ -0,0 +1,160 @@
+			parameters.pathfinding_algorithm = gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR;
+		} break;
+	}
+}
```
Again, is it worth delaying adding functions for different pathfinding until they are actually needed, because it may be that a different API is decided upon at that point (practical considerations etc)?

smix8 replied:
See discussion above with the PathfindingAlgorithm enum. I think it would be reasonable to keep this so ppl have the same access point to add new algorithm.

lawnjelly commented on modules/navigation/navigation_path_query_parameters_2d.cpp:
```diff
@@ -0,0 +1,160 @@
+void NavigationPathQueryParameters2D::set_target_position(const Vector2 &p_target_position) {
+	target_position_vec2 = p_target_position;
+	parameters.target_position = Vector3(p_target_position.x, 0.0, p_target_position.y);
+}
```
Do you need a `target_position_vec2`? Can you not just construct the `Vector2` as required from the `Vector3`?

smix8 replied:
This should no longer be required.

lawnjelly commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+		Callable::CallError callback_error;
+		// query callback object
+		query_callback_object->callp(query_callback_method, args, arg_count, callback_error);
+	}
```
Are there any race conditions here? Can the `query_callback_object` be deleted from another thread in between the time that you get it and call `callp`?

smix8 replied:
As far as I am aware of the Godot loop in the time window that the NavigationServer process this there is no deletion going on as all the user scripts frees() or queue_free() deletions of objects are already handled for the frame way before? Does it still require a lock in this case?

smix8 replied:
removed from this pr

DarkKilauea commented:
I was looking around other game engines and I have a new idea for the `query_path` API.

``` cpp

void query_path_async(const Ref<NavigationPathQueryParameters3D> &p_query_parameters, Callable p_query_result_callback);

```

The idea here is to put the callback into the method signature of `query_path` instead of placing on the result object that is pre-allocated.  Then, the caller would provide both the parameters and the callback to get the result.

This is also more similar to how the physics APIs work, though they are purely synchronous.  I'm a little torn on whether we should provide both a synchronous and asynchronous API, since I'm not sure all of the machinery for asynchronous path calculation is really worth it for most games.  Right now, my PC can do thousands of path calculations at >60 FPS with a synchronous API.

For simple games, they may actually get a performance bump by opting out of the async API and avoiding the overhead.

The sync API would look like:

``` cpp

Ref<NavigationPathQueryResult3D> query_path_sync(const Ref<NavigationPathQueryParameters3D> &p_query_parameters) const;

```

DarkKilauea commented:
I'd be really tempted to drop the asynchronous support here and push out a basic synchronous API similar to how the physics server works today.  We know this paradigm works for current Godot games and may help unblock this PR and give us time to figure out the asynchronous logic later.

We can always add support later with a new method.

lawnjelly commented:
> I was looking around other game engines and I have an idea for the `query_path` API.
> 
> ```c++
> void query_path_async(const Ref<NavigationPathQueryParameters3D> &p_query_parameters, Callable p_query_result_callback);
> ```
> 
> The idea here is to put the callback into the method signature of `query_path` instead of placing on the result object that is pre-allocated. Then, the caller would provide both the parameters and the callback to get the result.

Yes I actually agree here. I'm not sure I understand the whole sending the result object? Is this a misunderstanding from one of my earlier comments?

I agree with the API suggested by @DarkKilauea , bar the problem of callable not being present in 3.x (I don't know whether we are planning on keeping these in sync).

If necessary, we could drop async from this PR, it may be quicker for me or another contributor to add it as it requires some familiarity with threading.

smix8 commented:
Removed the callback and processqueue from the pr, basically only the navquery core left now.

DarkKilauea commented:
I believe we should move the query parameter and result objects next to the navigation server itself, instead of putting them into the module.

1. I want to be able to use them inside `NavigationAgent` without having to disable functionality is the module is disabled.  This is particularly needed to make working with navigation links easier and more intuitive.  Right now, there isn't a way to trigger when an agent enters a link and I'd like that capability to always be there.
2. We should define a minimum set of query features that is required for other navigation nodes to work.  It's ok if future server implementations add features, but they should all implement the base set.  Moving the types next to the server would allow other implementations to derive from our base parameters and result objects and then add their new features, with a cast used on the consumption side to gain access.

smix8 commented:
I agree but that module issue is not simply only for the navquery.

The current situation is that nearly every navigation pr that adds something new requires a lot of ifdef guards to even pass the recently added git check for Godot minimum build which is very annoying because the navigation module is in no state to be disabled as shown be issues like https://github.com/godotengine/godot/issues/36091 or the recent discussions on rocketchat (with no real solution) how to even properly disable such an intertwined module like navigation.

If the navigation module folder is intended to be only for the godot_navigation_server backend implementation that basically means we a) need a dummy server for navigation similar to physics and b) the entire stuff that has general use like the navigationmeshgenerator or the navquery stuff needs to be moved from the module navigation to a server navigation folder.

smix8 commented:
Changed the folder structure and moved the parameters / results from the godot navigation module folder to "servers/navigation" as they are considered a part of the NavigationServer that needs to be available for other classes or server implementations that replace the godot server module.

DarkKilauea commented:
I did some bench-marking to try to understand the performance difference between this PR and #65680.

Code:
``` gdscript
var nav_map_rid := get_world_2d().navigation_map;
var start_location := NavigationServer2D.map_get_closest_point(nav_map_rid, start.global_position);
var target_location := NavigationServer2D.map_get_closest_point(nav_map_rid, end.global_position);

var result := NavigationPathQueryResult2D.new();
var query := NavigationPathQueryParameters2D.new();
query.map = nav_map_rid;
query.start_position = start_location;
query.target_position = target_location;

for i in 10:
	var average_buffer := PackedInt32Array();
	
	var current_run_time = run_time_seconds;
	while(current_run_time >= 0.0):
		var cur_time := Time.get_ticks_usec();
		
		NavigationServer2D.query_path(query, result);
		
		var end_time := Time.get_ticks_usec();
		var duration := end_time - cur_time;
		var duration_in_sec := duration * 0.000001;
		
		current_run_time -= duration_in_sec;
		average_buffer.push_back(duration);
	
	var average_duration := 0;
	for dur in average_buffer:
		average_duration += dur;
	
	average_duration /= average_buffer.size();
	print("Average query time: %s usec" % average_duration);
```

Build Info:
```
> scons target=release_debug
scons: Reading SConscript files ...
Automatically detected platform: windows
Auto-detected 16 CPU cores available for build parallelism. Using 15 cores by default. You can override it with the -j argument.
Found MSVC version 14.3, arch x86_64
Building for platform "windows", architecture "x86_64", editor, target "release_debug".
```

Results:
```
Godot Engine v4.0.alpha.custom_build.489922a0e - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce RTX 2080

Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
Average query time: 20 usec
```

See #65680 for details about the testing done there, but the short version is that this PR is about 13.04% (3 usec) faster.

DarkKilauea commented on modules/navigation/nav_utils.h:
```diff
@@ -150,6 +150,25 @@ struct ClosestPointQueryResult {
+	PATHFINDING_ALGORITHM_ASTAR = 0,
+};
+
+struct PathQueryParameters {
```
Maybe these structs should be moved up into `NavigationServer3D` in order to use them to pass information up to that level, without having to depend on this module.  I noticed that the physics servers both do this with their queries.  It would mean adding a definition for a struct when we add GDExtension support, but that is something we can do.

smix8 replied:
moved to `servers/navigation/navigation_utilities.h`

DarkKilauea commented on servers/navigation/navigation_path_query_parameters_2d.h:
```diff
@@ -0,0 +1,76 @@
+class NavigationPathQueryParameters2D : public RefCounted {
+	GDCLASS(NavigationPathQueryParameters2D, RefCounted);
+
+	gd::PathQueryParameters parameters;
```
Alternatively, we could move this up to the server level, allowing us to continue to use it without another copy.

lawnjelly replied:
Agree that sticking the query and result structures in the core engine seems to make more sense than in a module.

smix8 replied:
moved to `servers/navigation/navigation_utilities.h` where the rest of the core parameters and results are that should not depend on the Godot navigation module.

DarkKilauea commented:
Sorry for the sudden comment spam, didn't notice that GitHub does not post your comments until you "finish your review".  After playing with this PR for a while, I'm happy with its overall shape.  I left a few comments for improvement, but these can be completed later.

DarkKilauea commented on servers/navigation_server_3d.h:
```diff
@@ -243,6 +248,11 @@ class NavigationServer3D : public Object {
 	virtual void process(real_t delta_time) = 0;
 
+	/// Returns a customized navigation path using a query parameters object
+	void query_path(const Ref<NavigationPathQueryParameters3D> &p_query_parameters, NavigationPathQueryResult3D *p_query_result) const;
```
```suggestion
	void query_path(const Ref<NavigationPathQueryParameters3D> &p_query_parameters, Ref<NavigationPathQueryResult3D> r_query_result) const;
```

DarkKilauea commented on servers/navigation_server_3d.cpp:
```diff
@@ -485,3 +487,12 @@ bool NavigationServer3D::get_debug_enabled() const {
+///////////////////////////////////////////////////////
+
+void NavigationServer3D::query_path(const Ref<NavigationPathQueryParameters3D> &p_query_parameters, NavigationPathQueryResult3D *p_query_result) const {
+	const gd::PathQueryResult _query_result = _query_path(p_query_parameters->get_parameters());
```
```suggestion
        ERR_FAIL_COND(!p_query_parameters.is_valid());
	ERR_FAIL_COND(!r_query_result.is_valid());

	const gd::PathQueryResult _query_result = _query_path(p_query_parameters->get_parameters());
```

DarkKilauea commented on servers/navigation_server_2d.cpp:
```diff
@@ -409,3 +411,17 @@ bool FORWARD_1_C(agent_is_map_changed, RID, p_agent, rid_to_rid);
+void NavigationServer2D::query_path(const Ref<NavigationPathQueryParameters2D> &p_query_parameters, NavigationPathQueryResult2D *p_query_result) const {
+	const gd::PathQueryResult _query_result = NavigationServer3D::get_singleton()->_query_path(p_query_parameters->get_parameters());
+
+	Vector<Vector2> path_vec2;
```
Use the `vector_v3_to_v2` function instead?

smix8 commented on servers/navigation_server_3d.h:
```diff
@@ -243,6 +251,17 @@ class NavigationServer3D : public Object {
 	/// Note: This function is not thread safe.
 	virtual void process(real_t delta_time) = 0;
 
+#ifdef MODULE_NAVIGATION_ENABLED
```
No longer the case with the objects moved to the server folder.

lawnjelly commented on doc/classes/NavigationPathQueryParameters3D.xml:
```diff
@@ -0,0 +1,39 @@
+	<members>
+		<member name="calculate_path_length" type="bool" setter="set_calculate_path_length" getter="get_calculate_path_length" default="false">
+			If [code]true[/code], the length of the path will be calculated as the sum of the distances between the path points.
+		</member>
```
Couldn't you just do this on demand? (while changing the name of the getter from `get_path_length()` to `find_path_length()` or `calculate_path_length()` to make it clear there is a calculation involved).

That would make this query parameter redundant.

smix8 replied:
Well the idea was to have any kind of path stat calculation as part of the query so the stats can be calculated in a threaded / async manner by the NavigationServer when the parameters for it is set true (so already on demand). Since the async was removed from this pr this part is now missing and it looks like a leftover that could be replaced with a static function anywhere.

smix8 replied:
I will remove it completely from this pr, maybe add as a static helper function later somewhere.

smix8 commented on modules/navigation/navigation_path_query_result_2d.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void NavigationPathQueryResult2D::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_path"), &NavigationPathQueryResult2D::get_path);
+	ClassDB::bind_method(D_METHOD("get_path_length"), &NavigationPathQueryResult2D::get_path_length);
```
Added the binds and properties.

smix8 commented on servers/navigation_server_3d.h:
```diff
@@ -33,14 +33,19 @@
 #include "core/object/class_db.h"
 #include "core/templates/rid.h"
+
+#include "modules/modules_enabled.gen.h"
```
True, there were 1-2 leftovers like this that are now removed.

smix8 commented:
Removed the path len calculation and the "optimize" bool parameter that switched between the two available postprocessing options in map_get_path(). Added path postprocessing enum as a replacement for the outdated "optimize" parameter.

DarkKilauea commented on servers/navigation_server_2d.cpp:
```diff
@@ -409,3 +411,9 @@ bool FORWARD_1_C(agent_is_map_changed, RID, p_agent, rid_to_rid);
 
 void FORWARD_1_C(free, RID, p_object, rid_to_rid);
+
+void NavigationServer2D::query_path(const Ref<NavigationPathQueryParameters2D> &p_query_parameters, Ref<NavigationPathQueryResult2D> p_query_result) const {
```
Guard against params and result being invalid refs?

DarkKilauea commented:
I'm happy with this PR.  It should be good to merge.
